libmamba/src/core/package_handling.cpp

```diff
@@ -443,46 +474,87 @@ namespace mamba
+
         archive_write_close(ext);
         archive_write_free(ext);
 
         fs::current_path(prev_path);
     }
 
+
+    static la_ssize_t file_read(struct archive* a, void* client_data, const void** buff)
+    {
+        struct conda_extract_context* mine = static_cast<conda_extract_context*>(client_data);
+        *buff = mine->buffer.data();
+
+        auto read = archive_read_data(mine->source, mine->buffer.data(), mine->buffer.size());
+        if (read < 0)
+        {
+            throw std::runtime_error(
+                fmt::format("Error reading from archive: {}", archive_error_string(mine->source)));
+        }
+        return read;
+    }
+
+    int archive_read_open_archive_entry(struct archive* a, conda_extract_context* ctx)
+    {
+        archive_clear_error(a);
+        archive_read_set_read_callback(a, file_read);
+        archive_read_set_callback_data(a, ctx);
+        return archive_read_open1(a);
+    }
+
+
     void extract_conda(const fs::u8path& file,
                        const fs::u8path& dest_dir,
                        const std::vector<std::string>& parts)
     {
-        TemporaryDirectory tdir;
-        extract_archive(file, tdir);
+        // open outer zip archive
+        struct archive* a = archive_read_new();
+        struct archive_entry* entry;
+
+        assert(archive_read_support_format_zip(a) == ARCHIVE_OK);
+
+        conda_extract_context extract_context(a);
 
-        auto fn = file.stem();
+        if (archive_read_open_filename(a, file.string().c_str(), extract_context.buffer.size())
+            != ARCHIVE_OK)
+        {
+            throw std::runtime_error(archive_error_string(a));
+        }
 
-        auto metadata_path = tdir.path() / "metadata.json";
-        if (fs::exists(metadata_path) && fs::file_size(metadata_path) != 0)
+        while (archive_read_next_header(a, &entry) == ARCHIVE_OK)
         {
-            const auto condafile_path = tdir.path() / "metadata.json";
-            std::ifstream condafile_meta(condafile_path.std_path());
-            nlohmann::json j;
-            condafile_meta >> j;
+            fs::u8path p(archive_entry_pathname(entry));
+            if (p.extension() == ".zst")
+            {
+                // extract zstd file
+                struct archive* inner = archive_read_new();
+                archive_read_support_filter_zstd(inner);
+                archive_read_support_format_tar(inner);
 
-            if (j.find("conda_pkg_format_version") != j.end())
+                archive_read_open_archive_entry(inner, &extract_context);
+                stream_extract_archive(inner, dest_dir);
+
+                archive_read_close(inner);
+                archive_read_free(inner);
+            }
+            else if (p.extension() == ".json")
             {
-                if (j["conda_pkg_format_version"] != 2)
+                std::size_t json_size = archive_entry_size(entry);
+                std::string json;
+                json.resize(json_size);
```

Comment:
This is equivalent to `std::string json(json_size, char{});` which is shorter.
However I'm not sure if that's exactly what you want here. `\0` is the end of a `const char*` but not the end of `std::string` so it might be confusing using that string afterwards. I suppose it's not a big deal if the next function do overwrites the whole thing, but make sure that the last `\0` is not part of the `std::string` otherwise there might be surprised for code using that one.

Reply:
👍